internal/mempool/v1/mempool.go

```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
+	if numTxs >= txmp.config.Size || wtx.Size()+txBytes > txmp.config.MaxTxsBytes {
 		return types.ErrMempoolIsFull{
 			NumTxs:      numTxs,
 			MaxTxs:      txmp.config.Size,
-			TxsBytes:    sizeBytes,
+			TxsBytes:    txBytes,
 			MaxTxsBytes: txmp.config.MaxTxsBytes,
 		}
 	}
 
 	return nil
 }
 
-func (txmp *TxMempool) insertTx(wtx *WrappedTx) {
-	txmp.txStore.SetTx(wtx)
-	txmp.priorityIndex.PushTx(wtx)
-	txmp.heightIndex.Insert(wtx)
-	txmp.timestampIndex.Insert(wtx)
-
-	// Insert the transaction into the gossip index and mark the reference to the
-	// linked-list element, which will be needed at a later point when the
-	// transaction is removed.
-	gossipEl := txmp.gossipIndex.PushBack(wtx)
-	wtx.gossipEl = gossipEl
-
-	atomic.AddInt64(&txmp.sizeBytes, int64(wtx.Size()))
-}
-
-func (txmp *TxMempool) removeTx(wtx *WrappedTx, removeFromCache bool) {
-	if txmp.txStore.IsTxRemoved(wtx.hash) {
-		return
-	}
-
-	txmp.txStore.RemoveTx(wtx)
-	txmp.priorityIndex.RemoveTx(wtx)
-	txmp.heightIndex.Remove(wtx)
-	txmp.timestampIndex.Remove(wtx)
-
-	// Remove the transaction from the gossip index and cleanup the linked-list
-	// element so it can be garbage collected.
-	txmp.gossipIndex.Remove(wtx.gossipEl)
-	wtx.gossipEl.DetachPrev()
-
-	atomic.AddInt64(&txmp.sizeBytes, int64(-wtx.Size()))
-
-	if removeFromCache {
-		txmp.cache.Remove(wtx.tx)
-	}
-}
-
-// purgeExpiredTxs removes all transactions that have exceeded their respective
-// height and/or time based TTLs from their respective indexes. Every expired
-// transaction will be removed from the mempool entirely, except for the cache.
+// purgeExpiredTxs removes all transactions from the mempool that have exceeded
+// their respective height or time-based limits as of the given blockHeight.
+// Transactions removed by this operation are not removed from the cache.
 //
-// NOTE: purgeExpiredTxs must only be called during TxMempool#Update in which
-// the caller has a write-lock on the mempool and so we can safely iterate over
-// the height and time based indexes.
+// The caller must hold txmp.mtx exclusively.
 func (txmp *TxMempool) purgeExpiredTxs(blockHeight int64) {
 	now := time.Now()
-	expiredTxs := make(map[types.TxKey]*WrappedTx)
-
-	if txmp.config.TTLNumBlocks > 0 {
-		purgeIdx := -1
-		for i, wtx := range txmp.heightIndex.txs {
-			if (blockHeight - wtx.height) > txmp.config.TTLNumBlocks {
-				expiredTxs[wtx.tx.Key()] = wtx
-				purgeIdx = i
-			} else {
-				// since the index is sorted, we know no other txs can be be purged
-				break
-			}
+	cur := txmp.txs.Front()
+	for cur != nil {
+		// N.B. Grab the next element first, since if we remove cur its successor
+		// will be invalidated.
+		next := cur.Next()
+
+		w := cur.Value.(*WrappedTx)
+		if txmp.config.TTLNumBlocks > 0 && (blockHeight-w.height) > txmp.config.TTLNumBlocks {
+			txmp.removeTxByElement(cur)
+		} else if txmp.config.TTLDuration > 0 && now.Sub(w.timestamp) > txmp.config.TTLDuration {
+			txmp.removeTxByElement(cur)
 		}
```

Comment:
Do we want to mark these transactions as evicted in our metrics?

Reply:
That's a good question. I've added that for now, but this seems like a case where we might want a separate metric for these (ExpiredTransactions, say).